egs/wsj/s5/utils/combine_ali_dirs.sh

```diff
@@ -0,0 +1,95 @@
+# Copyright 2016  Xiaohui Zhang  Apache 2.0.
+
+# This script operates on alignment directories, such as exp/tri4a_ali
+
+# Begin configuration section. 
+cmd=run.pl
+extra_files= # specify addtional files in 'src-data-dir' to merge, ex. "file1 file2 ..."
+ref_data_dir= # data directory to be used as reference for rearranging alignments
+num_jobs= # number of jobs used to split the reference data directory.
+# End configuration section.
+set -x
+echo "$0 $@"  # Print the command line for logging
+
+if [ -f path.sh ]; then . ./path.sh; fi
+. parse_options.sh || exit 1;
+
+if [[ $# -lt 2 || ! "$ref_data_dir" || ! "$num_jobs" ]]; then
+  echo "Usage: combine_ali_dirs.sh --ref-data-dir data/train --num-jobs 32 [--extra-files 'file1 file2'] <dest-ali-dir> <src-ali-dir1> <src-ali-dir2> ..."
+  echo "Note, files that don't appear in the first source dir will not be added even if they appear in later ones."
+  echo "Other than alignments, only files from the first src dir are copied."
+  exit 1
+fi
+
+dest=$1;
+shift;
+first_src=$1;
+
+rm -r $dest 2>/dev/null
+mkdir -p $dest;
+
+export LC_ALL=C
+
+for dir in $*; do
+  if [ ! -f $dir/ali.1.gz ]; then
+    echo "$0: check if alignments (ali.*.gz) are present in $dir."
+    exit 1;
+  fi
+done
+
+for dir in $*; do
+  for f in tree; do
+    diff $first_src/$f $dir/$f 1>/dev/null 2>&1
+    if [ $? -ne 0 ]; then
+      echo "$0: Cannot combine alignment directories with different $f files."
+    fi
+  done
+done
+
+for f in final.mdl tree cmvn_opts splice_opts num_jobs $extra_files; do
+  if [ ! -f $first_src/$f ]; then
+    echo "combine_ali_dir.sh: no such file $first_src/$f"
+    exit 1;
+  fi
+  cp $first_src/$f $dest/
+done
+
+scr_id=0
+temp_dir=$dest/temp
+mkdir -p $temp_dir
+echo "$0: dumping alignments in each source directory as single archive and index."
+for dir in $*; do
+  scr_id=$((scr_id + 1))
+  cur_num_jobs=$(cat $dir/num_jobs) || exit 1;
+  all_ids=$(seq -s, $cur_num_jobs)
+  $cmd $dir/log/copy_alignments.log \
+    copy-int-vector "ark:gunzip -c $dir/ali.{$all_ids}.gz|" \
+    ark,scp:$temp_dir/ali.$scr_id.ark,$temp_dir/ali.$scr_id.scp || exit 1;
+done
+cat $temp_dir/ali.*.scp | sort -m > $temp_dir/ali.scp || exit 1;
+
+echo "$0: splitting the ref_data_dir to get reference utt2spk for individual ali.JOB.gz files."
+utils/split_data.sh $ref_data_dir $num_jobs || exit 1;
+echo $num_jobs > $dest/num_jobs  || exit 1
+   
+echo "$0: spliting the alignments to appropriate chunks according to the reference utt2spk files."
+for i in `seq 1 $num_jobs`; do
+  awk '{print $1}' $ref_data_dir/split$num_jobs/$i/utt2spk | sort > $temp_dir/utt_subset.$i
+  utils/filter_scp.pl $temp_dir/utt_subset.$i $temp_dir/ali.scp | copy-int-vector scp:- "ark,t:|gzip -c >$dest/ali.$i.gz" || exit 1
+done   
+
+echo "$0: checking the alignment files generated have at least 90% of the utterances."
+for i in `seq 1 $num_jobs`; do
+  gunzip -c $dest/ali.$i.gz | awk '{print $1}' | sort > $temp_dir/ali_utt_list.$i  || exit 1;
```

Comment:
I think you can get this list efficiently using 
`utils/filter_scp.pl $temp_dir/utt_subset.$i $temp_dir/ali.scp | awk '{print $1}'`